aws-cpp-sdk-transfer/source/transfer/UploadFileRequest.cpp

```diff
@@ -89,52 +92,136 @@ m_getObjectRetries(0),
+UploadFileRequest(fileName, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>{}, s3Client, executor, createBucket, doConsistencyChecks)
+{
+}
+    
+UploadFileRequest::UploadFileRequest(const std::shared_ptr<Aws::IOStream>& fileStream,
+                                     const Aws::String& bucketName,
+                                     const Aws::String& keyName,
+                                     const Aws::String& contentType,
+                                     Aws::Map<Aws::String, Aws::String>&& metadata,
+                                     const std::shared_ptr<Aws::S3::S3Client>& s3Client,
+                                     const std::shared_ptr<Aws::Utils::Threading::BlockingExecutor>& executor,
+                                     bool createBucket,
+                                     bool doConsistencyChecks) :
+UploadFileRequest("", fileStream, bucketName, keyName, contentType, std::move(metadata), s3Client, executor, createBucket, doConsistencyChecks, true)
+{
+}
+    
+UploadFileRequest::UploadFileRequest(const std::shared_ptr<Aws::IOStream>& fileStream,
+                                     const Aws::String& bucketName,
+                                     const Aws::String& keyName,
+                                     const Aws::String& contentType,
+                                     const Aws::Map<Aws::String, Aws::String>& metadata,
+                                     const std::shared_ptr<Aws::S3::S3Client>& s3Client,
+                                     const std::shared_ptr<Aws::Utils::Threading::BlockingExecutor>& executor,
                                      bool createBucket,
                                      bool doConsistencyChecks) :
-UploadFileRequest(fileName, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>(metadata), s3Client, createBucket, doConsistencyChecks)
+UploadFileRequest(fileStream, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>(metadata), s3Client, executor, createBucket, doConsistencyChecks)
+{
+}
+    
+UploadFileRequest::UploadFileRequest(const std::shared_ptr<Aws::IOStream>& fileStream,
+                                     const Aws::String& bucketName,
+                                     const Aws::String& keyName,
+                                     const Aws::String& contentType,
+                                     const std::shared_ptr<Aws::S3::S3Client>& s3Client,
+                                     const std::shared_ptr<Aws::Utils::Threading::BlockingExecutor>& executor,
+                                     bool createBucket,
+                                     bool doConsistencyChecks) :
+UploadFileRequest(fileStream, bucketName, keyName, contentType, Aws::Map<Aws::String, Aws::String>{}, s3Client, executor, createBucket, doConsistencyChecks)
 {
 }
 
 UploadFileRequest::~UploadFileRequest()
 {
-    m_fileStream.close();
+    m_executor->WaitForCompletion();

```

Comment:
RESOLUTION: move WaitForCompletion into the TransferClient but as discussed, there is still an existing timing hole here where the request can be cancelled, the UploadFileRequest deleted, and whatever wait is implemented here will fail (e.g., waiter will wait on an object that has been removed with random results).
Not fixing this for now